File: p78.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> solution = {};
        int n = nums.size();
        for(unsigned int i = 0; i < pow(2, n); ++i) {
            vector<int> subset;
            // digits of the binary number i describe whether the element is in this current subset or not
            for(int j = 0; j < n; ++j) {
                int mask = 1 << j;
                // the digit is a '1'
                if((i&mask) > 0)
                    subset.push_back(nums[j]);
            }
            solution.push_back(subset);
        }
        return solution;
    }
};

File: p78.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []

        def solve(idx, curr):
            if idx > len(nums)-1:
                ans.append(curr[:])
                return
            curr.append(nums[idx])
            solve(idx+1, curr)
            curr.pop(-1)
            solve(idx+1, curr)

        solve(0, [])

        return ans
